File: Exercise2/CSharp/Exercise2_Solution/Exercise2_Solution/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

using OSIsoft.AF;
using OSIsoft.AF.Asset;
using OSIsoft.AF.UnitsOfMeasure;
using OSIsoft.AF.Time;
using OSIsoft.AF.Notification;
using OSIsoft.AF.EventFrame;
using OSIsoft.AF.Data;
using OSIsoft.AF.Search;

namespace Exercise2_Solution
{
    public partial class MainForm : Form
    {
        private enum Feature { Temperature, CloudCover, WindSpeed, Visibility, Pressure, Humidity }
        private enum DataMethod { RecordedValues, InterpolatedValues, Summary }

        private static IDictionary<Feature, string> _mapToUomName => new Dictionary<Feature, string>() { { Feature.CloudCover, "percent" }
                                                                                                       , { Feature.Humidity, "percent" }
                                                                                                       , { Feature.Pressure, "millibar" }
                                                                                                       , { Feature.Temperature, "degree Celsius" }
                                                                                                       , { Feature.Visibility, "kilometer" }
                                                                                                       , { Feature.WindSpeed, "kilometer per hour" }
                                                                                                       };

        private class MetaData : IEquatable<MetaData>
        {
            public MetaData()
            { }

            public MetaData(MetaData source)
            {
                Element = source.Element;
                FeatureText = source.FeatureText;
                EngUnit = source.EngUnit;
                DataMethod = source.DataMethod;
                StartTime = source.StartTime;
                EndTime = source.EndTime;
            }

            public AFElement Element { get; set;}
            public string FeatureText { get; set; }
            public UOM EngUnit { get; set; } = null;
            public DataMethod DataMethod { get; set; } = DataMethod.RecordedValues;
            public string StartTime { get; set; }
            public string EndTime { get; set; }

            public AFAttribute Attribute => Element?.Attributes[FeatureText];

            public bool IsTimeRangeValid => ParseToAFTime(StartTime).HasValue && ParseToAFTime(EndTime).HasValue;

            public AFTimeRange? TimeRange => IsTimeRangeValid ? (AFTimeRange?)(new AFTimeRange(StartTime, EndTime)) : null;

            public bool IsValid => (Attribute != null) && IsTimeRangeValid;

            public bool Equals(MetaData other)
            {
                if ((object)other == null)
                {
                    return false;
                }
                return (Attribute == other.Attribute)
                    && (DataMethod == other.DataMethod)
                    && (EngUnit == other.EngUnit)
                    && (TimeRange == other.TimeRange);
            }

            public bool IsEmpty => Equals(new MetaData());

            public MetaData Clone => new MetaData(this);
        }

        private MetaData Selected = new MetaData();
        private MetaData Retrieved = new MetaData();

        public MainForm()
        {
            InitializeComponent();

            //To have text align correctly in listbox, we will use a Fixed Width font
            lboxDataValues.Font = new Font("Consolas", 9, FontStyle.Regular);

            FillIntervalPicks();
            cboxInterval.Enabled = false;

            rbTemperature.Checked = true;
            Selected.FeatureText = rbTemperature.Text;
            FillUomPicks(Feature.Temperature);

            rbRecordedValues.Checked = true;
            Selected.DataMethod = DataMethod.RecordedValues;

            Selected.StartTime = tbStartTime.Text;
            Selected.EndTime = tbEndTime.Text;

            afDatabasePicker1.SystemPicker = piSystemPicker1;
            afTreeView1.AFRoot = afDatabasePicker1.AFDatabase;

            CheckAllButtons();
        }

        private PISystem AssetServer => piSystemPicker1.PISystem;
        private AFDatabase Database => afDatabasePicker1.AFDatabase;

        private void afDatabasePicker1_SelectionChange(object sender, OSIsoft.AF.UI.SelectionChangeEventArgs e)
        {
            afTreeView1.AFRoot = Database?.Elements;
            CheckAllButtons();
        }

        private void afTreeView1_AfterSelect(object sender, TreeViewEventArgs e)
        {
            Selected.Element = afTreeView1.AFSelection as AFElement;
            CheckAllButtons();
        }

        private void rbFeature_CheckChanged(object sender, EventArgs e)
        {
            var radioButton = sender as RadioButton;
            if (radioButton == null || !radioButton.Checked)
                return;

            // Continue for the checked radio button
            Selected.FeatureText = radioButton.Text;
            var feature = (Feature)GetByText(typeof(Feature), radioButton.Text);
            FillUomPicks(feature);

            CheckAllButtons();
        }

        private void FillUomPicks(Feature feature)
        {
            cboxUom.Items.Clear();
            var defaultUOM = AssetServer?.UOMDatabase.UOMs[_mapToUomName[feature]];
            if (defaultUOM != null)
            {
                cboxUom.Items.Add($"< default ( {defaultUOM.Abbreviation} ) >");
                cboxUom.SelectedIndex = 0;
                foreach (var uom in defaultUOM.Class.UOMs)
                {
                    cboxUom.Items.Add(uom);
                }
            }
        }

        private void rbMethod_CheckChanged(object sender, EventArgs e)
        {
            var radioButton = sender as RadioButton;
            if (radioButton == null || !radioButton.Checked)
            {
                return;
            }

            // Continue for the checked radio button
            Selected.DataMethod = (DataMethod)GetByText(typeof(DataMethod), radioButton.Text);
            cboxInterval.Enabled = (Selected.DataMethod == DataMethod.InterpolatedValues);
            CheckAllButtons();
        }

        
        private static object GetByText(Type enumType, string text) => Enum.Parse(enumType, RemoveBlanks(text));
        private static string RemoveBlanks(string text) => text.Replace(" ", string.Empty);


        // While this seems to be a likely candidate for an extension method,
        // there is a strong possibility that text could be null, which
        // doesn't work well with extension methods.
        public static AFTime? ParseToAFTime(string text)
        {
            try
            {
                var time = AFTime.MaxValue;
                // AFTime.TryParse
                // https://techsupport.osisoft.com/Documentation/PI-AF-SDK/Html/M_OSIsoft_AF_Time_AFTime_TryParse_2.htm
                // DateTime.TryParse
                // https://msdn.microsoft.com/en-us/library/ch92fbc1(v=vs.110).aspx
                if (AFTime.TryParse(text, out time))
                {
                    return time;
                }
            }
            catch
            { }
            return null;
        }
        private AFTime? GetStartTime() => ParseToAFTime(tbStartTime.Text);
        private AFTime? GetEndTime() => ParseToAFTime(tbEndTime.Text);

        private void TimeBox_KeyUp(object sender, KeyEventArgs e)
        {
            var textBox = sender as TextBox;
            TimeBoxValidate(textBox);
            if (textBox == tbStartTime)
            {
                Selected.StartTime = textBox.Text;
            }
            else if (textBox == tbEndTime)
            {
                Selected.EndTime = textBox.Text;
            }
            CheckAllButtons();
        }

        private void TimeBoxValidate(TextBox textBox)
        {
            if (textBox == null)
                return;

            if (ParseToAFTime(textBox.Text).HasValue)
            {
                textBox.BackColor = Color.White;
                textBox.ForeColor = Color.Black;
            }
            else
            {
                textBox.BackColor = Color.LightYellow;
                textBox.ForeColor = Color.Red;
            }
        }

        private void CheckAllButtons()
        {
            btnViewElement.Enabled = (Selected.Attribute != null);
            btnGetData.Enabled = Selected.IsValid;

            lblMetaInfo.ForeColor = (Selected.Equals(Retrieved) || Retrieved.IsEmpty) ? Color.Black : Color.Red;
        }

        private void btnGetData_Click(object sender, EventArgs e)
        {
            lboxDataValues.Items.Clear();

            var data = Selected.Attribute.Data;

            if (Selected.DataMethod == DataMethod.Summary)
            {
                var summaryDict = Selected.Attribute.Data.Summary(Selected.TimeRange.Value, AFSummaryTypes.All, AFCalculationBasis.TimeWeighted, AFTimestampCalculation.Auto);
                foreach (var summary in summaryDict)
                {
                    var value = summary.Value;
                    var uomAbbr = (value.UOM != null) ? $" {value.UOM.Abbreviation}" : "";
                    lboxDataValues.Items.Add($"{summary.Key,17}\t{value.Value}{uomAbbr}\t{value.Timestamp}");
                }
            }
            else
            {
                var values = new AFValues();

                if (Selected.DataMethod == DataMethod.InterpolatedValues)
                {
                    var interval = (TimeSpan)cboxInterval.SelectedItem;
                    values = data.InterpolatedValues(Selected.TimeRange.Value, new AFTimeSpan(interval), Selected.EngUnit, null, false);
                }
                else
                {
                    values = data.RecordedValues(Selected.TimeRange.Value, AFBoundaryType.Interpolated, Selected.EngUnit, null, false, 0);
                }

                foreach (var value in values)
                {
                    var uomAbbr = (value.UOM != null) ? $" {value.UOM.Abbreviation}" : "";
                    lboxDataValues.Items.Add($"{value.Timestamp.LocalTime}\t{value.Value:N1}{uomAbbr}");
                }
            }

            // Let's be nice and show some context regarding the data values that were retrieved
            var sb = new StringBuilder();
            sb.Append(Selected.Attribute);
            if (Selected.EngUnit != null)
                sb.Append($"; UOM={Selected.EngUnit.Abbreviation}");
            sb.Append($"; {Selected.DataMethod}");
            sb.Append($"; {Selected.TimeRange.Value}");
            lblMetaInfo.Text = sb.ToString();

            Retrieved = Selected.Clone;
            CheckAllButtons();
        }

        private void cboxUom_SelectedIndexChanged(object sender, EventArgs e)
        {
            var cbox = sender as ComboBox;
            if (cbox == null)
                return;
            if (cbox.SelectedIndex < 1)
            {
                Selected.EngUnit = null;
            }
            else
            {
                Selected.EngUnit = (UOM)cbox.SelectedItem;
            }

            CheckAllButtons();
        }

        private void piSystemPicker1_SelectionChange(object sender, OSIsoft.AF.UI.SelectionChangeEventArgs e)
        {
            if (AssetServer == null)
                return;
            var feature = (Feature)GetByText(typeof(Feature), Selected.FeatureText);
            FillUomPicks(feature);
        }

        private void btnViewElement_Click(object sender, EventArgs e)
        {
            var dialogForm = new AttributesForm(Selected.Element);
            dialogForm.ShowDialog(this);
        }

        private void FillIntervalPicks()
        {
            cboxInterval.Items.Clear();
            cboxInterval.Items.Add(TimeSpan.FromMinutes(5));
            cboxInterval.Items.Add(TimeSpan.FromMinutes(15));
            cboxInterval.Items.Add(TimeSpan.FromHours(1));
            cboxInterval.SelectedIndex = 2;
            cboxInterval.Items.Add(TimeSpan.FromHours(8));
            cboxInterval.Items.Add(TimeSpan.FromDays(1));
            cboxInterval.Items.Add(TimeSpan.FromDays(7));
        }
    }
}

File: Exercise2/VB.NET/Exercise2_Solution/Exercise2_Solution/MainForm.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports OSIsoft.AF
Imports OSIsoft.AF.Asset
Imports OSIsoft.AF.UnitsOfMeasure
Imports OSIsoft.AF.Time
Imports OSIsoft.AF.Notification
Imports OSIsoft.AF.EventFrame
Imports OSIsoft.AF.Data
Imports OSIsoft.AF.Search

Public Class MainForm
    Inherits Form

    Private Enum Feature
        Temperature
        CloudCover
        WindSpeed
        Visibility
        Pressure
        Humidity
    End Enum

    Private Enum DataMethod
        RecordedValues
        InterpolatedValues
        Summary
    End Enum

    Private Shared ReadOnly Property _mapToUomName As IDictionary(Of Feature, String)
        Get
            Dim map = New Dictionary(Of Feature, String)
            map.Add(Feature.CloudCover, "percent")
            map.Add(Feature.Humidity, "percent")
            map.Add(Feature.Pressure, "millibar")
            map.Add(Feature.Temperature, "degree Celsius")
            map.Add(Feature.Visibility, "kilometer")
            map.Add(Feature.WindSpeed, "kilometer per hour")
            Return map
        End Get
    End Property

    Private Class MetaData
        Implements IEquatable(Of MetaData)

        Public Sub New()
            Element = Nothing
            FeatureText = "Temperature"
            EngUnit = Nothing
            DataMethod = DataMethod.RecordedValues
            StartTime = "y"
            EndTime = "*"
        End Sub

        Public Sub New(ByVal source As MetaData)
            Element = source.Element
            FeatureText = source.FeatureText
            EngUnit = source.EngUnit
            DataMethod = source.DataMethod
            StartTime = source.StartTime
            EndTime = source.EndTime
        End Sub

        Public Property Element As AFElement
        Public Property FeatureText As String
        Public Property EngUnit As UOM
        Public Property DataMethod As DataMethod
        Public Property StartTime As String
        Public Property EndTime As String

        Public ReadOnly Property Attribute As AFAttribute
            Get
                If (Element Is Nothing) OrElse String.IsNullOrWhiteSpace(FeatureText) Then
                    Return Nothing
                End If
                Return Element.Attributes(FeatureText)
            End Get
        End Property


        Public ReadOnly Property IsTimeRangeValid As Boolean
            Get
                Return ParseToAFTime(StartTime).HasValue AndAlso ParseToAFTime(EndTime).HasValue
            End Get
        End Property


        Public ReadOnly Property TimeRange As AFTimeRange?
            Get
                If IsTimeRangeValid Then
                    Return New AFTimeRange(StartTime, EndTime)
                End If
                Return Nothing
            End Get
        End Property

        Public ReadOnly Property IsValid As Boolean
            Get
                Return (Attribute IsNot Nothing) AndAlso IsTimeRangeValid
            End Get
        End Property

        Public Shadows Function Equals(ByVal other As MetaData) As Boolean Implements IEquatable(Of MetaData).Equals
            If (CType(other, Object) Is Nothing) Then
                Return False
            End If

            Return ((Me.Attribute = other.Attribute) _
                        AndAlso ((Me.DataMethod = other.DataMethod) _
                        AndAlso ((Me.EngUnit = other.EngUnit) _
                        AndAlso (Me.TimeRange = other.TimeRange))))
        End Function

        Public ReadOnly Property IsEmpty As Boolean
            Get
                Return Equals(New MetaData())
            End Get
        End Property

        Public Function Clone() As MetaData
            Return New MetaData(Me)
        End Function

        Public Overrides Function ToString() As String
            ' Let's be nice and show some context regarding the data values that were retrieved
            Dim sb = New StringBuilder
            sb.Append(Attribute)
            If (EngUnit IsNot Nothing) Then
                sb.Append($"; UOM={EngUnit.Abbreviation}")
            End If
            sb.Append($"; {DataMethod}")
            sb.Append($"; {TimeRange.Value}")
            Return sb.ToString()
        End Function
    End Class

    Private Selected As MetaData = New MetaData
    Private Const InitialMetaDataText As String = "Data Values"

    Public Sub New()
        InitializeComponent()

        lblMetaInfo.Text = InitialMetaDataText

        'To have text align correctly in listbox, we will use a Fixed Width font
        lboxDataValues.Font = New Font("Consolas", 9, FontStyle.Regular)

        FillIntervalPicks()
        cboxInterval.Enabled = False

        rbTemperature.Checked = True
        Selected.FeatureText = rbTemperature.Text
        FillUomPicks(Feature.Temperature)

        rbRecordedValues.Checked = True
        Selected.DataMethod = DataMethod.RecordedValues

        Selected.StartTime = tbStartTime.Text
        Selected.EndTime = tbEndTime.Text

        afDatabasePicker1.SystemPicker = piSystemPicker1
        afTreeView1.AFRoot = afDatabasePicker1.AFDatabase

        CheckAllButtons()
    End Sub

    Private ReadOnly Property AssetServer As PISystem
        Get
            Return piSystemPicker1.PISystem
        End Get
    End Property

    Private ReadOnly Property Database As AFDatabase
        Get
            Return afDatabasePicker1.AFDatabase
        End Get
    End Property

    Private Sub afDatabasePicker1_SelectionChange(ByVal sender As Object, ByVal e As OSIsoft.AF.UI.SelectionChangeEventArgs) Handles afDatabasePicker1.SelectionChange
        afTreeView1.AFRoot = Database?.Elements
        Me.CheckAllButtons()
    End Sub

    Private Sub afTreeView1_AfterSelect(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles afTreeView1.AfterSelect
        Selected.Element = TryCast(afTreeView1.AFSelection, AFElement)
        CheckAllButtons()
    End Sub

    Private Sub rbFeature_CheckChanged(ByVal sender As Object, ByVal e As EventArgs) Handles rbWindSpeed.CheckedChanged, rbVisibility.CheckedChanged, rbTemperature.CheckedChanged, rbPressure.CheckedChanged, rbHumidity.CheckedChanged, rbCloudCover.CheckedChanged
        Dim radioButton = TryCast(sender, RadioButton)
        If (radioButton Is Nothing OrElse Not radioButton.Checked) Then
            Return
        End If

        ' Continue for the checked radio button
        Selected.FeatureText = radioButton.Text
        Dim feature = CType(GetByText(GetType(Feature), radioButton.Text), Feature)
        FillUomPicks(feature)
        CheckAllButtons()
    End Sub

    Private Sub FillUomPicks(ByVal feature As Feature)
        cboxUom.Items.Clear()
        Dim defaultUOM = AssetServer?.UOMDatabase.UOMs(_mapToUomName(feature))
        If (defaultUOM IsNot Nothing) Then
            cboxUom.Items.Add($"< default ( {defaultUOM.Abbreviation} ) >")
            cboxUom.SelectedIndex = 0
            For Each uom In defaultUOM.Class.UOMs
                cboxUom.Items.Add(uom)
            Next
        End If

    End Sub

    Private Sub rbMethod_CheckChanged(ByVal sender As Object, ByVal e As EventArgs) Handles rbSummary.CheckedChanged, rbRecordedValues.CheckedChanged, rbInterpolatedValues.CheckedChanged
        Dim radioButton = TryCast(sender, RadioButton)
        If (radioButton Is Nothing OrElse Not radioButton.Checked) Then
            Return
        End If

        ' Continue for the checked radio button
        Selected.DataMethod = CType(GetByText(GetType(DataMethod), radioButton.Text), DataMethod)
        cboxInterval.Enabled = (Me.Selected.DataMethod = DataMethod.InterpolatedValues)
        Me.CheckAllButtons()
    End Sub

    Private Shared Function GetByText(ByVal enumType As Type, ByVal text As String) As Object
        If (String.IsNullOrWhiteSpace(text)) Then Return 0
        Return System.Enum.Parse(enumType, RemoveBlanks(text))
    End Function

    Private Shared Function RemoveBlanks(ByVal text As String) As String
        Return text.Replace(" ", String.Empty)
    End Function

    ' While this seems to be a likely candidate for an extension method,
    ' there is a strong possibility that text could be null, which
    ' doesn't work well with extension methods.
    Public Shared Function ParseToAFTime(ByVal text As String) As AFTime?
        If String.IsNullOrWhiteSpace(text) Then Return Nothing
        Try
            Dim time = AFTime.MaxValue
            ' AFTime.TryParse
            ' https://techsupport.osisoft.com/Documentation/PI-AF-SDK/Html/M_OSIsoft_AF_Time_AFTime_TryParse_2.htm
            ' DateTime.TryParse
            ' https://msdn.microsoft.com/en-us/library/ch92fbc1(v=vs.110).aspx
            If AFTime.TryParse(text, time) Then
                Return time
            End If
        Catch
        End Try

        Return Nothing
    End Function

    Private Function GetStartTime() As AFTime?
        Return ParseToAFTime(tbStartTime.Text)
    End Function

    Private Function GetEndTime() As AFTime?
        Return ParseToAFTime(tbEndTime.Text)
    End Function

    Private Sub TimeBox_KeyUp(ByVal sender As Object, ByVal e As KeyEventArgs) Handles tbStartTime.KeyUp, tbEndTime.KeyUp
        Dim textBox = TryCast(sender, TextBox)
        TimeBoxValidate(textBox)
        If (textBox Is tbStartTime) Then
            Selected.StartTime = textBox.Text
        ElseIf (textBox Is tbEndTime) Then
            Selected.EndTime = textBox.Text
        End If

        CheckAllButtons()
    End Sub

    Private Sub TimeBoxValidate(ByVal textBox As TextBox)
        If (textBox Is Nothing) Then
            Return
        End If

        If ParseToAFTime(textBox.Text).HasValue Then
            textBox.BackColor = Color.White
            textBox.ForeColor = Color.Black
        Else
            textBox.BackColor = Color.LightYellow
            textBox.ForeColor = Color.Red
        End If

    End Sub

    Private Sub CheckAllButtons()
        btnViewElement.Enabled = (Selected.Attribute IsNot Nothing)
        btnGetData.Enabled = Selected.IsValid

        If (lblMetaInfo.Text = InitialMetaDataText) OrElse (lblMetaInfo.Text = Selected.ToString()) Then
            lblMetaInfo.ForeColor = Color.Black
        Else
            lblMetaInfo.ForeColor = Color.Red
        End If
    End Sub

    Private Sub btnGetData_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnGetData.Click
        lboxDataValues.Items.Clear()

        Dim data = Selected.Attribute.Data

        If (Selected.DataMethod = DataMethod.Summary) Then
            Dim summaryDict = data.Summary(Selected.TimeRange.Value, AFSummaryTypes.All, AFCalculationBasis.TimeWeighted, AFTimestampCalculation.Auto)
            For Each summary In summaryDict
                Dim value = summary.Value
                Dim uomAbbr As String = ""
                If (value.UOM IsNot Nothing) Then
                    uomAbbr = $" {value.UOM.Abbreviation}"
                End If
                lboxDataValues.Items.Add($"{summary.Key,17}{vbTab}{value.Value}{uomAbbr}{vbTab}{value.Timestamp}")
            Next
        Else
            Dim values = New AFValues
            If (Selected.DataMethod = DataMethod.InterpolatedValues) Then
                Dim interval = CType(cboxInterval.SelectedItem, TimeSpan)
                values = data.InterpolatedValues(Selected.TimeRange.Value, New AFTimeSpan(interval), Selected.EngUnit, Nothing, False)
            Else
                values = data.RecordedValues(Selected.TimeRange.Value, AFBoundaryType.Interpolated, Selected.EngUnit, Nothing, False, 0)
            End If

            For Each value In values
                Dim uomAbbr As String = ""
                If (value.UOM IsNot Nothing) Then
                    uomAbbr = $" {value.UOM.Abbreviation}"
                End If
                lboxDataValues.Items.Add($"{value.Timestamp.LocalTime}{vbTab}{value.Value:N3}{uomAbbr}")
            Next
        End If

        lblMetaInfo.Text = Selected.ToString()
        CheckAllButtons()
    End Sub

    Private Sub cboxUom_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cboxUom.SelectedIndexChanged
        Dim cbox = TryCast(sender, ComboBox)
        If (cbox Is Nothing) Then
            Return
        End If

        If (cbox.SelectedIndex < 1) Then
            Selected.EngUnit = Nothing
        Else
            Selected.EngUnit = CType(cbox.SelectedItem, UOM)
        End If

        CheckAllButtons()
    End Sub

    Private Sub piSystemPicker1_SelectionChange(ByVal sender As Object, ByVal e As OSIsoft.AF.UI.SelectionChangeEventArgs) Handles piSystemPicker1.SelectionChange
        If (AssetServer Is Nothing) Then
            Return
        End If

        Dim feature = CType(GetByText(GetType(Feature), Selected.FeatureText), Feature)
        Me.FillUomPicks(feature)
    End Sub

    Private Sub btnViewElement_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnViewElement.Click
        Dim dialogForm = New AttributesForm(Selected.Element)
        dialogForm.ShowDialog(Me)
    End Sub

    Private Sub FillIntervalPicks()
        cboxInterval.Items.Clear
        cboxInterval.Items.Add(TimeSpan.FromMinutes(5))
        cboxInterval.Items.Add(TimeSpan.FromMinutes(15))
        cboxInterval.Items.Add(TimeSpan.FromHours(1))
        cboxInterval.SelectedIndex = 2
        cboxInterval.Items.Add(TimeSpan.FromHours(8))
        cboxInterval.Items.Add(TimeSpan.FromDays(1))
        cboxInterval.Items.Add(TimeSpan.FromDays(7))
    End Sub
End Class
